torchqdynamics/mesolve.py

```diff
@@ -35,44 +50,69 @@ def mesolve(
     # define the QSolver
     if isinstance(solver, Rouchon):
         if solver.order == 1:
-            qsolver = MERouchon1(H, jump_ops, solver)
+            qsolver = MERouchon1(H_batched, jump_ops, solver)
         elif solver.order == 1.5:
-            qsolver = MERouchon1_5(H, jump_ops, solver)
+            qsolver = MERouchon1_5(H_batched, jump_ops, solver)
         elif solver.order == 2:
-            qsolver = MERouchon2(H, jump_ops, solver)
+            qsolver = MERouchon2(H_batched, jump_ops, solver)
     else:
         raise NotImplementedError
 
     # compute the result
-    return odeint(qsolver, rho0, t_save, exp_ops, save_states, gradient_alg)
+    y_save, exp_save = odeint(qsolver, y0, t_save, exp_ops, save_states, gradient_alg)
+
+    # restore correct batching
+    if rho0.dim() == 2:
+        y_save = y_save.squeeze(1)
+        exp_save = exp_save.squeeze(1)
+    if H.dim() == 2:
+        y_save = y_save.squeeze(0)
+        exp_save = exp_save.squeeze(0)
+
+    return y_save, exp_save
 
 
 class MERouchon(AdjointQSolver):
     def __init__(
         self, H: TimeDependentOperator, jump_ops: List[torch.Tensor],
         solver_options: SolverOption
     ):
-        self.H = H
-        self.jump_ops = jump_ops
-        self.jumpdag_ops = jump_ops.adjoint()
-        self.sum_nojump = (self.jumpdag_ops @ self.jump_ops).sum(dim=0)
-        self.I = torch.eye(H(0).shape[-1]).to(H(0))
+        # Args:
+        #     H: (b_H, n, n)
+
+        # we convert H and jump_ops to compatible sizes, i.e. tensors of size
+        # (a, b, n, n) where a is b_H or 1 and b is len(jump_ops) or 1
```

Comment:
`# (b_H, len(jump_ops), n, n)`
No need to introduce (a,b) notations I think. It's confusing here since you don't use it later.